src/main/scala/translating/ILtoIL.scala

```diff
@@ -0,0 +1,279 @@
+    node
+  }
+
+  override def visitProcedure(node: Procedure): Procedure = {
+    program ++= "Procedure(" + procedureIdentifier(node) + ", "
+    indentLevel += 1
+  
+      program ++= "in("
+    for (i <- node.in.indices) {
+      node.in(i) = visitParameter(node.in(i))
+      if (i != node.in.size - 1)
+        program ++= ", "
+    }
+    program ++= "), "
+    program ++= "out("
+    for (i <- node.out.indices) {
+      node.out(i) = visitParameter(node.out(i))
+      if (i != node.out.size - 1)
+        program ++= ", "
+    }
+    program ++= "), "
+    program ++= "blocks(\n"
+    for (i <- node.blocks.indices) {
+      node.blocks(i) = visitBlock(node.blocks(i))
+    }
+    program ++= "))\n"
+    indentLevel -= 1
+    node
+  }
+
+  override def visitParameter(node: Parameter): Parameter = {
+    program ++= "Parameter("
+    node.value = visitRegister(node.value)
+    program ++= ")"
+    node
+  }
+
+  override def visitProgram(node: Program): Program = {
+    for (i <- node.procedures.indices) {
+      val updatedProcedure = visitProcedure(node.procedures(i))
+      val targetProcedure = node.procedures(i)
+      if (targetProcedure == node.mainProcedure) {
+        node.mainProcedure = updatedProcedure
+      }
+      node.procedures(i) = updatedProcedure
+    }
+    node
+  }
+
+  override def visitExtract(node: Extract): Expr = {
+    program ++= "Extract("
+    node.body = visitExpr(node.body)
+    program ++= ")"
+    node
+  }
+
+  override def visitRepeat(node: Repeat): Expr = {
+    program ++= "Repeat("
+    node.body = visitExpr(node.body)
+    program ++= ")"
+    node
+  }
+
+  override def visitZeroExtend(node: ZeroExtend): Expr = {
+    program ++= "ZeroExtend("
+    node.body = visitExpr(node.body)
+    program ++= ")"
+    node
+  }
+
+  override def visitSignExtend(node: SignExtend): Expr = {
+    program ++= "SignExtend("
+    node.body = visitExpr(node.body)
+    program ++= ")"
+    node
+  }
+
+  override def visitUnaryExpr(node: UnaryExpr): Expr = {
+    program ++= "UnaryExpr("
+    node.arg = visitExpr(node.arg)
+    program ++= ")"
+    node
+  }
+
+  override def visitBinaryExpr(node: BinaryExpr): Expr = {
+    program ++= "BinaryExpr("
+    node.arg1 = visitExpr(node.arg1)
+    program ++= ", "
+    node.arg2 = visitExpr(node.arg2)
+    program ++= ")"
+    node
+  }
+
+  override def visitMemoryStore(node: MemoryStore): MemoryStore = {
+    program ++= "MemoryStore("
+    visitMemory(node.mem)
+    program ++= "["
+    node.index = visitExpr(node.index)
+    program ++= "] := "
+    node.value = visitExpr(node.value)
+    program ++= ")"
+    node
+  }
+
+  override def visitMemoryLoad(node: MemoryLoad): Expr = {
+    program ++= "MemoryLoad("
+    node.mem = visitMemory(node.mem)
+    program ++= ", ["
+    node.index = visitExpr(node.index)
+    program ++= "])"
+    node
+  }
+
+  override def visitMemory(node: Memory): Memory = {
+    program ++= "Memory("
+    program ++= node.toString()
```

Comment:
This results in outputting `Memory(Memory(` since Memory's `toString` also includes `Memory(`